some fixes for mwdt toolchain

IRISZZW opened the pull request:
- define __INT*_C() and __UINT*_C() for mwdt toolchain

    our minimal C library makes an alias of UINT*_C() to be __UINT*_C() and INT*_C() to __INT*_C().
    However, in mwdt, these are not defined by default, so define them ourselves.
    this commit reverts: https://github.com/zephyrproject-rtos/zephyr/commit/b1adc462be6258329ea94bced96e4457a0c4b5cb

- use ATOMIC_OPERATIONS_BUILTIN for mwdt toolchain

    ATOMIC_OPERATIONS_BUILTIN issue (internal jira number: P10019563-43273) has been fixed in 
    new relasese MWDT 2021.03. We can use builtin atomic.
    we have similar fix for xcc: https://github.com/zephyrproject-rtos/zephyr/commit/5ef8db8ac989e683fb1b51572b00074780f82f0a

- add arcmwdt toolchain support in hsdk.yaml, then we can
use arcmwdt for hsdk board with twister tool.

IRISZZW committed:
libc: minimal: mwdt: define __INT*_C() and __UINT*_C()

Our minimal C library makes an alias of UINT*_C() to
be __UINT*_C() and INT*_C() to __INT*_C(). However,
in mwdt, these are not defined by default, so define
them ourselves. We have similar fix for xcc: #31962

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

abrodkin commented on arch/arc/Kconfig:
```diff
@@ -78,6 +75,7 @@ config CPU_EM6
 config CPU_HS3X
 	bool
 	select CPU_ARCHS
+	select HAS_FLASH_LOAD_OFFSET
```
How this hunk is related to atomics?

IRISZZW replied:
I wonder why and when I add this too. removed.

abrodkin reviewed (changes requested):
1. Please refer to the commit you're basically reverting: https://github.com/zephyrproject-rtos/zephyr/commit/b1adc462be6258329ea94bced96e4457a0c4b5cb
2. Please refer to the similar fix for xcc: https://github.com/zephyrproject-rtos/zephyr/commit/5ef8db8ac989e683fb1b51572b00074780f82f0a

ruuddw commented:
Instead of adding the __INT*_C() and __UINT*_C() to a Metaware header, wouldn't it be better to add them to the minimal C lib stdint.h file? That already has the definitions for GNU.
Maybe something to discuss in the toolchain WG? Btw, MetaWare's stdint.h already defines the INT*C(). I would expect that the stdint.h would always come from the toolchain, instead of the Zephyr minimal C lib.

IRISZZW committed:
arch: archs: using ATOMIC_OPERATIONS_BUILTIN

ATOMIC_OPERATIONS_BUILTIN issue (internal jira number: P10019563-43273)
has been fixed in new relasese MWDT 2021.03. We can use builtin atomic.
this commit reverts PR: #28528

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>

evgeniy-paltsev commented on arch/arc/Kconfig:
```diff
@@ -18,10 +18,7 @@ config CPU_ARCEM
-	# toolchain, so choosing ATOMIC_OPERATIONS_C instead.
-	select ATOMIC_OPERATIONS_C if "$(ZEPHYR_TOOLCHAIN_VARIANT)" = "arcmwdt"
-	select ATOMIC_OPERATIONS_BUILTIN if "$(ZEPHYR_TOOLCHAIN_VARIANT)" != "arcmwdt"
+	select ATOMIC_OPERATIONS_BUILTIN
```
We have builtin atomic fix starting from MWDT 2021.03, so we need to check that we have appropriate MWDT toolchain.

@IRISZZW could you please add MWDT version check and don't allow to use MWDT version < 2021.03. The check can be placed in `cmake/compiler/arcmwdt/target.cmake` where we do compiler discovery.

evgeniy-paltsev replied:
PS: I'm totally OK if it is done as a follow-up pull request.

IRISZZW replied:
I will do it in another pull request.

IRISZZW commented:
> Instead of adding the __INT*_C() and __UINT*_C() to a Metaware header, wouldn't it be better to add them to the minimal C lib stdint.h file? That already has the definitions for GNU.
> Maybe something to discuss in the toolchain WG? Btw, MetaWare's stdint.h already defines the INT*C(). I would expect that the stdint.h would always come from the toolchain, instead of the Zephyr minimal C lib.

@ruuddw there is no feedback from other guys, maybe we can keep it in our own `mwdt.h` head file to let this fixes merged before release.

ruuddw commented:
>  there is no feedback from other guys, maybe we can keep it in our own `mwdt.h` head file to let this fixes merged before release.

If we can later rework of this PR, I think it makes sense to check that now. There is still some time. Which other guys are you referring to? Did you explicitly ask for feedback? Did you check if an implementation in the minimal C lib stdint.h would be feasible?

IRISZZW commented:
> > there is no feedback from other guys, maybe we can keep it in our own `mwdt.h` head file to let this fixes merged before release.
> 
> If we can later rework of this PR, I think it makes sense to check that now. There is still some time. Which other guys are you referring to? Did you explicitly ask for feedback? Did you check if an implementation in the minimal C lib stdint.h would be feasible?

I havn't explicity ask for feedback. I have checked thar mwdt has these macros in `stdint.h` file. when we use zephyr minimal C lib, it also has `stdint.h` file. we will use the `stdin.h` file from zephyr minimal C lib, not from mwdt.

IRISZZW commented:
@ruuddw I have moved __INT*_C() and __UINT*_C() to minimal C lib stdint.h file.

dcpleung commented on boards/arc/hsdk/hsdk.yaml:
```diff
@@ -4,7 +4,7 @@ type: mcu
 toolchain:
   - zephyr
-  - xtools
+  - arcmwdt
```
Please do not remove xtools. It allows using custom built GCC toolchain, and is useful when testing builds from Zephyr's sdk-ng repo.

IRISZZW replied:
OK, added.

IRISZZW committed:
board: hsdk: add arcmwdt toolchain support in hsdk.yaml

add arcmwdt toolchain support in hsdk.yaml, then we can
use arcmwdt for hsdk board with twister tool.

Signed-off-by: Watson Zeng <zhiwei@synopsys.com>
